# Single-Cell RNA Sequencing (scRNA-seq) Analysis
Overview
Single-cell RNA sequencing (scRNA-seq) is a powerful technique that enables the measurement of gene expression at the level of individual cells, offering a higher resolution view of cellular diversity compared to traditional bulk RNA sequencing. While bulk RNA-seq measures the average gene expression across a population of cells, scRNA-seq isolates and analyzes RNA from each cell separately, allowing researchers to uncover subtle differences in gene expression that are often masked in pooled samples. This ability to capture cellular heterogeneity is particularly valuable for identifying rare cell types, understanding complex developmental processes, and exploring disease mechanisms. The typical workflow involves isolating individual cells, capturing their RNA, converting it into complementary DNA (cDNA), and sequencing it to generate vast amounts of data. Advanced bioinformatics tools are then used to process, analyze, and interpret the resulting data, often revealing previously unrecognized cell types, states, and regulatory networks. Although scRNA-seq provides rich insights into cellular function, it also presents challenges, such as data sparsity, technical noise, and high costs. Nevertheless, its applications are vast, ranging from cancer research and immune system studies to developmental biology and drug discovery, making it a transformative tool in modern genomics.

# Projects
# Clustering of Peripheral Blood Mononuclear Cells (PBMCs) 
<img width="346" alt="Screenshot 2024-12-20 at 8 12 48â€¯PM" src="https://github.com/user-attachments/assets/a85818b6-dbb1-4d83-8472-1f33993e935d" />

